sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -258,6 +258,24 @@ object FileSourceStrategy extends Strategy with PredicateHelper with Logging {
       val outputAttributes = readDataColumns ++ generatedMetadataColumns ++
         partitionColumns ++ constantMetadataColumns
 
+
+      // The metadata attribute references in the filters also have to be categorized as either
+      // constant or generated metadata attributes. Only data filters can contain metadata filters.
+      def categorizeFileSourceMetadataAttributesInFilters(
+          filters: Seq[Expression]): Seq[Expression] =
+        filters.map { filter =>
+          filter.transform {
+            case attr: AttributeReference if FileSourceMetadataAttribute.unapply(attr).isDefined =>
+              if (attr.dataType.asInstanceOf[StructType].fieldNames
+                .forall(fieldName => constantMetadataColumns.exists(fieldName == _.name))) {
```

Comment:
Seems we flatten the metadata column in the scan node, shall we flatten the metadata cols in the filter as well?

Reply:
That is a really good idea, I did settle for something along those lines: I am rebinding the references to the flattened columns. This way the filters do again reference columns that are actually part of the schema